Source/WebGPU/WGSL/GlobalVariableRewriter.cpp

```diff
@@ -0,0 +1,281 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "GlobalVariableRewriter.h"
+
+#include "AST.h"
+#include "ASTVisitor.h"
+#include "CallGraph.h"
+#include "WGSL.h"
+#include <wtf/HashMap.h>
+#include <wtf/HashSet.h>
+
+namespace WGSL {
+
+class RewriteGlobalVariables : public AST::Visitor {
+public:
+    RewriteGlobalVariables(CallGraph& callGraph, const HashMap<String, PipelineLayout>& pipelineLayouts)
+        : AST::Visitor()
+        , m_callGraph(callGraph)
+    {
+        UNUSED_PARAM(pipelineLayouts);
+    }
+
+    void run();
+
+    void visit(AST::FunctionDecl&) override;
+    void visit(AST::VariableDecl&) override;
+    void visit(AST::IdentifierExpression&) override;
+
+private:
+    template<typename Value>
+    using IndexMap = HashMap<unsigned, Value, WTF::IntHash<unsigned>, WTF::UnsignedWithZeroKeyHashTraits<unsigned>>;
+
+    using IndexSet = HashSet<unsigned, WTF::IntHash<unsigned>, WTF::UnsignedWithZeroKeyHashTraits<unsigned>>;
+
+    struct Global {
+        struct Resource {
+            unsigned m_group;
+            unsigned m_binding;
+        };
+
+        std::optional<Resource> m_resource;
+        AST::VariableDecl* m_declaration;
+    };
+
+    static String argumentBufferParameterName(unsigned group);
+    static String argumentBufferStructName(unsigned group);
+
+    void def(const String&);
+    Global* read(const String&);
+
+    void collectGlobals();
+    void insertStructs();
+    void visitEntryPoint(AST::FunctionDecl&);
+    IndexSet requiredGroups();
+    void insertParameters(AST::FunctionDecl&, const IndexSet& requiredGroups);
+
+    CallGraph& m_callGraph;
+    HashMap<String, Global> m_globals;
+    IndexMap<IndexMap<Global*>> m_groupBindingMap;
+    HashSet<String> m_defs;
+    HashSet<String> m_reads;
+};
+
+void RewriteGlobalVariables::run()
+{
+    collectGlobals();
+    insertStructs();
+    for (auto& entryPoint : m_callGraph.entrypoints())
+        visitEntryPoint(entryPoint.m_function);
+    m_callGraph.ast().globalVars().clear();
+}
+
+void RewriteGlobalVariables::visit(AST::FunctionDecl& functionDecl)
+{
+    // FIXME: visit callee once we have any
+
+    for (auto& parameter : functionDecl.parameters())
+        def(parameter.name());
```

Comment:
Usually we use full words

Reply:
Will rename. We should probably also rename all the `[...]Decl` nodes.